### PR TITLE
Interval token store

### PR DESCRIPTION
**Pull request by josojo**
This is the first draft for the depositing and withdrawing infrastructure of the stablecoin-converter.

Generally, the infrastructure handles deposits and withdrawals depending on a `stateIndex`. Deposits will be pending for exactly one `stateIndex` and then they will be credited right with the next transaction touching the account.
Withdraws need to be requested in a previous `prev_stateIndex`, and then they can be withdrawn anytime, as long current `stateIndex` >  `prev_stateIndex`.

Later on, it will be used like this:
```
contract Stablecoin-converter is IntervalTokenStore {

 // modify balances via:
substractBalance(address user, address token, uint amount)
addBalance(address user, address token, uint amount)

```

At some later stage, we need to add SAFEERC20 transfer that deals with the different ERC20 standards out here...

### CODE REVIEW

File: contracts/IntervalTokenStore.sol [hunk 1]
```diff
@@ -0,0 +1,142 @@
+// IntervalTokenStore stores tokens for applications, which have discrete states increasing with time
```
Comment (fleupold) [inline contracts/IntervalTokenStore.sol]: This should go right above the contract definition. I'd also reword it a bit. Maybe call it IntervalTokenLocker or EpochTokenLocker instead of Store?

```
Allows to deposit token which become credited in the next epoch and allows to request a token withdraw which becomes claimable after the current epoch expired.
```

File: contracts/IntervalTokenStore.sol [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+    struct BalanceState {
+        uint256 balance;
+        PendingFlux pendingDeposits;
+        PendingFlux pendingWithdraws;
+    }
+
+    struct PendingFlux {
+        uint256 amount;
+        uint256 stateIndex;  // deposits will be processed for any currentStateIndex > stateIndex
```
Comment (fleupold) [inline contracts/IntervalTokenStore.sol]: maybe "will be credited". Also maybe lift these comments up to pendingDeposits/pendingWithdraw field on BalanceState

File: contracts/IntervalTokenStore.sol [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+                              // withdraws are allowed for any currentStateIndex > stateIndex
+    }
+
+    uint256 public currentStateIndex=0;
```
Comment (fleupold) [inline contracts/IntervalTokenStore.sol]: nit: space between =

File: contracts/IntervalTokenStore.sol [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+    function updateAndGetBalance(address user, address token) public returns(uint256) {
+        updateDepositsBalance(user, token);
```
Comment (fleupold) [inline contracts/IntervalTokenStore.sol]: What's the need to update the balance here? Can't we just update whenever we deposit and leave this function view?

File: contracts/IntervalTokenStore.sol [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        balanceStates[msg.sender][token].pendingDeposits.stateIndex = currentStateIndex;
+        emit Deposit(msg.sender, token, amount, currentStateIndex);
+    }
+
+    function withdrawRequest(address token, uint amount) public {
+        balanceStates[msg.sender][token].pendingWithdraws = PendingFlux({ amount: amount, stateIndex: currentStateIndex });
+        emit WithdrawRequest(msg.sender, token, amount, currentStateIndex);
+    }
+
+    function withdraw(address token, uint amount) public {
```
Comment (fleupold) [inline contracts/IntervalTokenStore.sol]: Why do we have to specify an amount here? I'd assume it just gives me whatever is available.

File: contracts/IntervalTokenStore.sol [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+            ERC20(token).transferFrom(msg.sender, address(this), amount),
+            "Tokentransfer for deposit was not successful"
+        );
+        balanceStates[msg.sender][token].pendingDeposits.amount += amount;
+        balanceStates[msg.sender][token].pendingDeposits.stateIndex = currentStateIndex;
+        emit Deposit(msg.sender, token, amount, currentStateIndex);
+    }
+
+    function withdrawRequest(address token, uint amount) public {
+        balanceStates[msg.sender][token].pendingWithdraws = PendingFlux({ amount: amount, stateIndex: currentStateIndex });
```
Comment (fleupold) [inline contracts/IntervalTokenStore.sol]: Is it intended that a previous withdraw would just be overridden?

File: contracts/test/IntervalTokenStoreTestInterface.sol [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+
+contract IntervalTokenStoreTestInterface is IntervalTokenStore {
+
+
+    function increaseStateIndex() public {
+        currentStateIndex++;
+    }
+
+    function addBalanceTest(address user, address token, uint amount) public {
```
Comment (fleupold) [inline contracts/test/IntervalTokenStoreTestInterface.sol]: Why are these needed?

File: contracts/IntervalTokenStore.sol [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+        balanceStates[msg.sender][token].balance -= amount;
+        delete balanceStates[msg.sender][token].pendingWithdraws;
+
+        ERC20(token).transfer(msg.sender, amount);
+        emit Withdraw(msg.sender, token, amount);
+    }
+
+    function updateDepositsBalance(address user, address token) public {
+        if (balanceStates[user][token].pendingDeposits.stateIndex < currentStateIndex) {
+            balanceStates[user][token].balance += balanceStates[user][token].pendingDeposits.amount;
```
Comment (fleupold) [inline contracts/IntervalTokenStore.sol]: use SafeMath

File: contracts/IntervalTokenStore.sol [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+            balanceStates[msg.sender][token].pendingWithdraws.amount >= amount,
+            "registered withdraw-amount was not sufficient"
+        );
+
+        require(
+            balanceStates[msg.sender][token].balance >= amount,
+            "balances not sufficient"
+        );
+
+        balanceStates[msg.sender][token].balance -= amount;
```
Comment (fleupold) [inline contracts/IntervalTokenStore.sol]: SafeMath

File: contracts/IntervalTokenStore.sol [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+        return balance;
+    }
+
+    function deposit(address token, uint amount) public {
+        updateDepositsBalance(msg.sender, token);
+        require(
+            ERC20(token).transferFrom(msg.sender, address(this), amount),
+            "Tokentransfer for deposit was not successful"
+        );
+        balanceStates[msg.sender][token].pendingDeposits.amount += amount;
```
Comment (fleupold) [inline contracts/IntervalTokenStore.sol]: SafeMath

File: contracts/IntervalTokenStore.sol [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+        emit Deposit(msg.sender, token, amount, currentStateIndex);
+    }
+
+    function withdrawRequest(address token, uint amount) public {
+        balanceStates[msg.sender][token].pendingWithdraws = PendingFlux({ amount: amount, stateIndex: currentStateIndex });
+        emit WithdrawRequest(msg.sender, token, amount, currentStateIndex);
+    }
+
+    function withdraw(address token, uint amount) public {
+        updateDepositsBalance(msg.sender, token);
```
Comment (fleupold) [inline contracts/IntervalTokenStore.sol]: Maybe comment that this is necessary in order to avoid negative balance (deposits might have to be credited in order to withdraw).

File: contracts/IntervalTokenStore.sol [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+        require(
+            ERC20(token).transferFrom(msg.sender, address(this), amount),
+            "Tokentransfer for deposit was not successful"
+        );
+        balanceStates[msg.sender][token].pendingDeposits.amount += amount;
+        balanceStates[msg.sender][token].pendingDeposits.stateIndex = currentStateIndex;
+        emit Deposit(msg.sender, token, amount, currentStateIndex);
+    }
+
+    function withdrawRequest(address token, uint amount) public {
```
Comment (fleupold) [inline contracts/IntervalTokenStore.sol]: nit: requestWithdraw sounds more imperative

**Review by fleupold — COMMENTED** [review]
Looks good overall. Yet, there is a lot of things going on in this PR so asking some questions inline.

File: contracts/IntervalTokenStore.sol (hunk 5 above)
Comment (josojo) [reply contracts/IntervalTokenStore.sol]: Well, yeah usually people want to have all of it. But if someone changes his mind, he is able to withdraw less now.
Since it does not add more work, I would keep it.

File: contracts/IntervalTokenStore.sol (hunk 8 above)
Comment (josojo) [reply contracts/IntervalTokenStore.sol]: Technically, the safemaths are not needed here, but you are right, better safe than sorry later.

File: contracts/IntervalTokenStore.sol (hunk 4 above)
Comment (josojo) [reply contracts/IntervalTokenStore.sol]: Later on, when we trade, we want to make sure that we are always working with the latest balance available. Hence, we enforce any reading through the interfaces:
`updateAndGetBalance, addBalance, substractBalance` and all of them make sure that the balances are updated before any other operation will happen.

This is my current thinking, maybe, you are aware of  a cleaner way to do it. :)

File: contracts/EpochTokenLocker.sol [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+    struct PendingFlux {
+        uint256 amount;
+        uint256 stateIndex;
+    }
+
+    uint256 public currentStateIndex = 0;
+
+    function updateAndGetBalance(address user, address token) public returns(uint256) {
+        updateDepositsBalance(user, token);
+        uint balance = balanceStates[user][token].balance;
```
Comment (anxolin) [inline contracts/EpochTokenLocker.sol]: Would it be more gas efficient to save balanceStates[user][token] in a local var.
This would apply also to the other functions

File: contracts/IntervalTokenStore.sol (hunk 5 above)
Comment (fleupold) [reply contracts/IntervalTokenStore.sol]: It makes the interface more confusing in my eyes. This is why I'd suggest to not allow this kind of weird behaviour in the first place (since there is no way to cancel a withdraw, there is no point in not withdrawing all of it even if one decided one doesn't need it)

File: contracts/IntervalTokenStore.sol (hunk 4 above)
Comment (fleupold) [reply contracts/IntervalTokenStore.sol]: I agree that all writes to the balance should update pending deposits (and there should be an interface that people are forced to use). However, I feel like we should allow people to read the balances without requiring a tx (e.g. the driver will want to query all balances before it starts the solver without sending n transactions).

File: contracts/EpochTokenLocker.sol [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+        if (balanceStates[user][token].pendingWithdraws.stateIndex < currentStateIndex) {
+            balance -= Math.min(balanceStates[user][token].pendingWithdraws.amount, balance);
+        }
+        return balance;
+    }
+
+    function deposit(address token, uint amount) public {
+        updateDepositsBalance(msg.sender, token);
```
Comment (fleupold) [inline contracts/EpochTokenLocker.sol:51]: Who decides which tokens can be deposited? Do we want to limit this here or only on the trading level?

File: contracts/EpochTokenLocker.sol [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+    );
+
+    event Withdraw(
+        address user,
+        address token,
+        uint amount
+    );
+
+    // User => Token => BalanceState
+    mapping(address => mapping(address => BalanceState)) private balanceStates;
```
Comment (fleupold) [inline contracts/EpochTokenLocker.sol:35]: At some point we will need a way to iterate over all users in the system.

**Review by fleupold — COMMENTED** [review]
Two more thoughts

File: contracts/EpochTokenLocker.sol [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    function getBalance(address user, address token) public view returns(uint) {
+        return balanceStates[user][token].balance;
+    }
+
+    /**
+     * internal functions
+     */
+    function addBalance(address user, address token, uint amount) internal {
```
Comment (anxolin) [inline contracts/EpochTokenLocker.sol:130]: I don't see this function used in the smart contract. I see that is used only for testing purposes.

If you need it only for testing, is not better to do one of the following?:
- define it just in the test contracts
- Use the mocking library
- Do a setup where u do the actual deposit of tokens

File: test/epoch_token_locker.js [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+
+const truffleAssert = require("truffle-assertions")
+
+
+contract("EpochTokenLocker", async (accounts) => {
+  const [user_1] = accounts
+
+  describe("deposit", () => {
+    it("processes a deposit and stores it in the pendingDeposits", async () => {
+      const instance = await EpochTokenLocker.new()
```
Comment (anxolin) [inline test/epoch_token_locker.js]: Not super fan of calling something instance. Maybe `epochTokenLocker`, since it's an instance of `EpochTokenLocker` contract.

File: test/epoch_token_locker.js [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+      const ERC20 = await MockContract.new()
+      await ERC20.givenAnyReturnBool(true)
+      await instance.deposit(ERC20.address, 100)
+      await instance.deposit(ERC20.address, 100)
+
+      assert.equal(await instance.getPendingDepositAmount(user_1, ERC20.address), 200)
+      assert.equal(await instance.getPendingDepositBatchNumber(user_1, ERC20.address), 0)
+    })
+
+    it("does not add two deposits, if they are not deposited during same stateIndex", async () => {
```
Comment (anxolin) [inline test/epoch_token_locker.js]: Wouldn't make more sense to test:

"A deposit, consolidates any pending deposit"

And ensure that the user has:
* balance: 100
* pending deposit: 100

Also, instead of using the same amount 100 + 100, consider to use 100 and 50 for example, to make sure they are saved in the right place

**Review by anxolin — APPROVED** [review]
Looks good to me :) 
I added some small cosmetic comments

Nice start Alex!

File: contracts/EpochTokenLocker.sol (hunk 15 above)
Comment (anxolin) [reply contracts/EpochTokenLocker.sol:35]: In this implementation it would be only by using the events.
Is not enough? Or u mean it should have something to make it easier to iterate just with the smart contract:

```
mapping(address => uint) private userIds;
mapping(address => uint) private tokenIds;
address[] users;
address[] tokens;
```

File: contracts/EpochTokenLocker.sol (hunk 14 above)
Comment (anxolin) [reply contracts/EpochTokenLocker.sol:51]: I think it's good that the concerns are separated. It can be a specialization of this contract. One reason is because the list of tokens could have different strategies/implementation on how this whitelisted tokens are maintained

File: contracts/IntervalTokenStore.sol (hunk 4 above)
Comment (anxolin) [reply contracts/IntervalTokenStore.sol]: I don't have super strong opinion here, but I think you don't need a tx for getting the balance.

On the one hand, there's already this view function that shows the unupdated balance:
```
function getBalance(address user, address token) public view returns(uint) {
        return balanceStates[user][token].balance;
    }
```

And in the other, if you do a `call` instead of a `tx`, you would get the "real balance" taking the pending amounts into account, but it doesn't modify the blockchain or cost anything the transactions.

I also see it a bit strange a method that is called `updateAndGetBalance`, @fleupold does it sound better if it's just renamed to `updateBalance` (it's normal that a updateBalance returns the value of the updated balance).

File: contracts/EpochTokenLocker.sol (hunk 15 above)
Comment (fleupold) [reply contracts/EpochTokenLocker.sol:35]: yes that exactly. I'd like to avoid having to maintain a second layer infrastructure.

File: test/epoch_token_locker.js (hunk 17 above)
Comment (fleupold) [reply test/epoch_token_locker.js]: we usually call the unit under test instance, but I'm also fine with being more explicit.

File: contracts/IntervalTokenStore.sol (hunk 4 above)
Comment (fleupold) [reply contracts/IntervalTokenStore.sol]: Yes, I think that name would be better. I'd still argue there should be a view function just returning the current balance (without writing it).

File: contracts/IntervalTokenStore.sol (hunk 5 above)
Comment (josojo) [reply contracts/IntervalTokenStore.sol]: Got your point. You are right. We can not allow cancelations of withdraws.

File: contracts/EpochTokenLocker.sol (hunk 16 above)
Comment (josojo) [reply contracts/EpochTokenLocker.sol:130]: The function defines an interface, which will be used later. At least to my current thinking

File: contracts/EpochTokenLocker.sol (hunk 13 above)
Comment (josojo) [reply contracts/EpochTokenLocker.sol]: To my understanding, the optimizer is already smart enough to deal with these kinds of things.
I am wondering, probably loading balaceStates[user][token] would also require loading 5 variables, instead of just loading the 3 ones we really need. (If the optimizer is not smart enough)...

File: contracts/IntervalTokenStore.sol (hunk 4 above)
Comment (josojo) [reply contracts/IntervalTokenStore.sol]: Hm...
I feel that EpochTokenLocker should have a defined interface with a unique method to `get` the current balance. And due to this unusual design pattern, we need to update and read the balance at the same time. That is also why I would prefer to have both getBalance and update in the name. 
If we have the only `updateBalance`, people might wonder where they can get the balance from in the first place. I think it would be less intuitive.

 >I'd still argue there should be a view function just returning the current balance (without writing it).

I am happy to add another function. This could be useful for other smart contracts, which want to act conditionally on the exchange and don't have the ability to do a .call. Is this the purpose you have in mind, Felix?

File: contracts/EpochTokenLocker.sol (hunk 16 above)
Comment (fleupold) [reply contracts/EpochTokenLocker.sol:130]: Should it then really be defined as an interface? At the moment it's just a one-off internal functions with no call-sites (except for some contract used for testing). Do we need it for this PR?

File: contracts/IntervalTokenStore.sol (hunk 4 above)
Comment (fleupold) [reply contracts/IntervalTokenStore.sol]: > And due to this unusual design pattern, we need to update and read the balance at the same time.

We don't really have to update the balance whenever we want to read it. We could just evaluate the pending deposit and leave it there. We only really need to update when we want to mutate the balance, which is already a tx (I think your suggestion makes sense to only allow that via an interface we control).

However, I think that the reading and writing interface should be separate. Therefore I advertise for a view function that allows to read the balance of the current user that doesn't have side effects.

Comment (anxolin) [reply contracts/IntervalTokenStore.sol]: I think it's similar as banks, when you receive a transfer and is not settled, the bank shows both the total balance and the available balance (at least my bank :P )

Also, sometimes you use your credit card, and they don't take the money right away, but they remove it from your available balance while the operation is confirmed.

I find it ok to show both, it makes sense to me too Felix suggestion of having two view fn:
* `getTotalBalance`
* `getAvailableBalance`

Additionally we would need the `updateBalance` or `applyPendingBalance`

File: contracts/EpochTokenLocker.sol (hunk 13 above)
Comment (anxolin) [reply contracts/EpochTokenLocker.sol]: I didn't understand last part.
Yep, it would make sense that the optimizer do that. In any case, I think it would be also more readable imo.

File: test/epoch_token_locker.js (hunk 17 above)
Comment (anxolin) [reply test/epoch_token_locker.js]: Ok. I think is a fair naming convention. I've seen it called "subject", instance is ok too.
I like it either way, so we can continue as you've been doing

File: contracts/IntervalTokenStore.sol (hunk 4 above)
Comment (josojo) [reply contracts/IntervalTokenStore.sol]: Felix and I decided that actually only the getTotalBalance is needed. Other value is not of relevance.

and I implemented the updateBalance for sure

Comment (anxolin) [reply contracts/IntervalTokenStore.sol]: But why? 

I'm reviewing again, and the `withdraw` is a bit different than the deposit. When you deposit, you've already sent the tokens, so they are yours already (they are just not available).

If you send a withdraw, the actual withdraw is not done. And even if you consume your balance and then try to withdraw, you won't be able to do so. 

For this reason, I still think it's relevant something like:
* **Total balance**: `balance + any deposits`. Shows the tokens that the contract owes you.
* **Available balance**: `balance + applicable deposits`. It's your usable balance.



Additional, I think it would be useful to show also the:
* **Current balance**: balance

A dapp could show always the total balance, and the available only if they don't match.

Current balance could be use to decide if you can update the balance (when `current <> available`)

File: contracts/EpochTokenLocker.sol [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+            balanceStates[msg.sender][token].pendingWithdraws.stateIndex < currentStateIndex,
+            "withdraw was not registered previously"
+        );
+
+        uint amount = Math.min(
+            balanceStates[msg.sender][token].balance,
+            balanceStates[msg.sender][token].pendingWithdraws.amount
+        );
+
+        balanceStates[msg.sender][token].balance = balanceStates[msg.sender][token].balance.sub(amount);
```
Comment (anxolin) [inline contracts/EpochTokenLocker.sol:80]: Lines `92-96` I think is best if they are executed only if `amount > 0`.
I guess is not super important, but we avoid emitting events of `amount = 0`

Comment (josojo) [reply contracts/EpochTokenLocker.sol:80]: Yeah, for me it is hard to foresee exactly how the events are used and so it is hard to code it properly.
If we want to be able to calculate everything from events, we might want to have the event even for amounts = 0. Because then, we need to know from somewhere, that the pendingWithdraws was deleted in order to calculate the balance correctly in the future. And currently the only way to know it is from the event.

I would like to leave it as it is and come back to it later.

File: contracts/EpochTokenLocker.sol (hunk 13 above)
Comment (josojo) [reply contracts/EpochTokenLocker.sol]: balaceStates[user][token] is a struct with 5 variables. If I would define it inside the smart contract, it might have the impact that all 5 variables need to be read (5*500 gas) instead of just the 3 that are needed. 
But I am not sure, how good the optimizer is. Maybe he changes the code so that only 3 are read.

File: contracts/EpochTokenLocker.sol (hunk 15 above)
Comment (josojo) [reply contracts/EpochTokenLocker.sol:35]: This will be implemented later in a separate PR. We want to use a set library, which is not yet implemented.

**Review by fleupold — APPROVED** [review]

File: contracts/EpochTokenLocker.sol (hunk 13 above)
Comment (anxolin) [reply contracts/EpochTokenLocker.sol]: Not sure, I would think that saving it in a intermediary variable would only save the reference to the start of the struct, so my intuition would say no variable would be loaded (we just save the pointer)

The we do the same reads using the pointer as the start position, and the fields would tell where to start reading.

It's usually a good practice to do this small refactoring for every programming language (specially for readability, but also a bit more performant), I would be surprised if in solidity is bad. Not sure how to make sure, it's not super important, but I'm sure this is something that is repeated in all the contracts since we will be working a lot with arrays of structs.

File: contracts/EpochTokenLocker.sol [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+        uint256 stateIndex;
+    }
+
+    uint256 public currentStateIndex = 0;
+
+    function deposit(address token, uint amount) public {
+        updateDepositsBalance(msg.sender, token);
+        require(
+            ERC20(token).transferFrom(msg.sender, address(this), amount),
+            "Tokentransfer for deposit was not successful"
```
Comment (bh2smith) [inline contracts/EpochTokenLocker.sol:54]: nit: `Tokentransfer` is two words.

File: contracts/EpochTokenLocker.sol [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+
+    /**
+     * internal functions
+     */
+    function addBalance(address user, address token, uint amount) internal {
+        updateDepositsBalance(msg.sender, token);
+        balanceStates[user][token].balance = balanceStates[user][token].balance.add(amount);
+    }
+
+    function substractBalance(address user, address token, uint amount) internal {
```
Comment (bh2smith) [inline contracts/EpochTokenLocker.sol:135]: "substract" is basically french for "subtract"

File: contracts/EpochTokenLocker.sol [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+        );
+        balanceStates[msg.sender][token].pendingDeposits.amount = balanceStates[msg.sender][token].pendingDeposits.amount
+            .add(amount);
+        balanceStates[msg.sender][token].pendingDeposits.stateIndex = currentStateIndex;
+        emit Deposit(msg.sender, token, amount, currentStateIndex);
+    }
+
+    function requestWithdraw(address token, uint amount) public {
+        balanceStates[msg.sender][token].pendingWithdraws = PendingFlux({ amount: amount, stateIndex: currentStateIndex });
+        emit WithdrawRequest(msg.sender, token, amount, currentStateIndex);
```
Comment (bh2smith) [inline contracts/EpochTokenLocker.sol:64]: Are we using events here in the same way we were for the Snapp contracts? Is this purely for the convenience of unit testing? I guess my question is, are these events intended to be used a reference for a future transaction?

File: package.json [hunk 23]
```diff
@@ -47,6 +47,7 @@
     "fast-memoize": "^2.5.1",
     "merkletreejs": "0.0.22",
     "mongodb": "^3.1.13",
+    "@gnosis.pm/mock-contract": "^3.0.7",
```
Comment (bh2smith) [inline package.json:50]: Sweet!

File: test/epoch_token_locker.js [hunk 24]
```diff
@@ -0,0 +1,223 @@
...
+      const ERC20 = await MockContract.new()
+      await ERC20.givenAnyReturnBool(true)
+      await epochTokenLocker.deposit(ERC20.address, 100)
+      await epochTokenLocker.deposit(ERC20.address, 100)
+
+      assert.equal(await epochTokenLocker.getPendingDepositAmount(user_1, ERC20.address), 200)
+      assert.equal(await epochTokenLocker.getPendingDepositBatchNumber(user_1, ERC20.address), 0)
+    })
+
+    it("does not consolidates two deposits, if they are not deposited during same stateIndex", async () => {
```
Comment (bh2smith) [inline test/epoch_token_locker.js:41]: nit: "consolidate"

File: test/epoch_token_locker.js [hunk 25]
```diff
@@ -0,0 +1,223 @@
...
+
+      await epochTokenLocker.addBalanceTest(user_1, ERC20.address, 100)
+
+      assert.equal(await epochTokenLocker.getBalance(user_1, ERC20.address), 100)
+    })
+  })
+  describe("substractBalance", () => {  
+    it("modifies the balance by substracting", async () => {
+      const epochTokenLocker = await EpochTokenLockerTestInterface.new()
+      const ERC20 = await MockContract.new()
```
Comment (bh2smith) [inline test/epoch_token_locker.js:209]: Interesting! Should be be doing this instead of actually using `MintableERC20` in the Snapp contract tests? Or is this just an "easy-way-out"

**Review by bh2smith — COMMENTED** [review]
Looks great. I see its already merged. Just a couple generic questions and some typos.
